### PR TITLE
pci: vfio: Preserve prefetchable flag on BAR

### PR DESCRIPTION
**Pull request by rbradford**
- Set prefetchable bit with VFIO devices
- pci: vfio: Preserve prefetchable state of BAR across
- HACK: Run VFIO worker on this PR

### CODE REVIEW

**Comment by sboeuf** [comment]
@rbradford could you please merge the two commits (yours and @sdake's) together since the correct fix is the combination of the two.

**Comment by rbradford** [comment]
@sboeuf Yeh, I will do that if it works.

**Comment by sboeuf** [comment]
> @sboeuf Yeh, I will do that if it works.

Cool, but honestly the fix is legit, no matter if it fixes @sdake's issue or not.

**Comment by sdake** [comment]
@rbradford would you mind adding a co-authored-by: Steven Dake <sdake@lambdal.com>? I did spend weeks looking at this problem 👍 

TY

**Comment by rbradford** [comment]
You did the bulk of the work on finding the issue so you should have the credit. Are you happy with this PR and does it solve your issue?

**Review by sdake — APPROVED** [review]
I rebased this work onto PR #https://github.com/cloud-hypervisor/cloud-hypervisor/pull/4443. The results look good. For example, 8 A100 GPUs startup, and the results of nvidia-smi topo -m are:
```
root@ubuntu:~# nvidia-smi topo -m
	GPU0	GPU1	GPU2	GPU3	GPU4	GPU5	GPU6	GPU7	CPU Affinity	NUMA Affinity
GPU0	 X 	NV12	NV12	NV12	NV12	NV12	NV12	NV12	0-15		N/A
GPU1	NV12	 X 	NV12	NV12	NV12	NV12	NV12	NV12	0-15		N/A
GPU2	NV12	NV12	 X 	NV12	NV12	NV12	NV12	NV12	0-15		N/A
GPU3	NV12	NV12	NV12	 X 	NV12	NV12	NV12	NV12	0-15		N/A
GPU4	NV12	NV12	NV12	NV12	 X 	NV12	NV12	NV12	0-15		N/A
GPU5	NV12	NV12	NV12	NV12	NV12	 X 	NV12	NV12	0-15		N/A
GPU6	NV12	NV12	NV12	NV12	NV12	NV12	 X 	NV12	0-15		N/A
GPU7	NV12	NV12	NV12	NV12	NV12	NV12	NV12	 X 	0-15		N/A
````

NV12 = good. I also took a look at lspci, and the BARs are correct for mem32 and mem64.

I then dropped the HEAD~1 and HEAD~3, only keeping this PR on top of my latest HEAD.

with the same results:
```
root@ubuntu:~# nvidia-smi topo -m
	GPU0	GPU1	GPU2	GPU3	GPU4	GPU5	GPU6	GPU7	CPU Affinity	NUMA Affinity
GPU0	 X 	NV12	NV12	NV12	NV12	NV12	NV12	NV12	0-15		N/A
GPU1	NV12	 X 	NV12	NV12	NV12	NV12	NV12	NV12	0-15		N/A
GPU2	NV12	NV12	 X 	NV12	NV12	NV12	NV12	NV12	0-15		N/A
GPU3	NV12	NV12	NV12	 X 	NV12	NV12	NV12	NV12	0-15		N/A
GPU4	NV12	NV12	NV12	NV12	 X 	NV12	NV12	NV12	0-15		N/A
GPU5	NV12	NV12	NV12	NV12	NV12	 X 	NV12	NV12	0-15		N/A
GPU6	NV12	NV12	NV12	NV12	NV12	NV12	 X 	NV12	0-15		N/A
GPU7	NV12	NV12	NV12	NV12	NV12	NV12	NV12	 X 	0-15		N/A
```

The good, or bad news is without the allowlist patch, the full PCI config space is available including SRIOV devices.

This PR LGTM.

File: pci/src/vfio.rs [hunk 1]
```diff
@@ -1505,6 +1515,8 @@ const PCI_CONFIG_CAPABILITY_OFFSET: u32 = 0x34;
 const PCI_CONFIG_IO_BAR: u32 = 0x1;
 // 64-bit memory bar flag.
 const PCI_CONFIG_MEMORY_BAR_64BIT: u32 = 0x4;
+// Prefetchable BAR bit
```
[the remark this answers is not in the document]
Comment (rbradford) [reply pci/src/vfio.rs:1518]: For flag values it's much easier to read like this.

**Comment by rbradford** [comment]
Undrafted. VFIO CI passed.

**Review by sboeuf — APPROVED** [review]

**Pull request merged**